File: ios/main/StallionSlotManager.m
//
//  StallionSlotManager.m
//  react-native-stallion
//
//  Created by Jasbir Singh Shergill on 29/01/25.
//

#import "StallionSlotManager.h"
#import "StallionStateManager.h"
#import "StallionMetaConstants.h"
#import "StallionEventHandler.h"
#import "StallionObjConstants.h"
#import "StallionFileManager.h"

@implementation StallionSlotManager

+ (void)rollbackProdWithAutoRollback:(BOOL)isAutoRollback errorString:(NSString *)errorString {
    StallionStateManager *stateManager = [StallionStateManager sharedInstance];
    StallionMeta *meta = stateManager.stallionMeta;
    SlotStates currentProdSlot = meta.currentProdSlot;
    NSString *stableReleaseHash = meta.prodStableHash ?: @"";
    NSString *newReleaseHash = meta.prodNewHash ?: @"";

    switch (currentProdSlot) {
        case SlotStateNewSlot:
            [meta setProdNewHash:@""];
            if (stableReleaseHash.length == 0) {
                [meta setCurrentProdSlot:SlotStateDefaultSlot];
            } else {
                [meta setCurrentProdSlot:SlotStateStableSlot];
            }
            [self emitRollbackEvent:isAutoRollback rolledBackReleaseHash:newReleaseHash errorString:errorString];
            [stateManager syncStallionMeta];
            break;

        case SlotStateStableSlot:
            [meta setProdStableHash:@""];
            [meta setCurrentProdSlot:SlotStateDefaultSlot];
            [self emitRollbackEvent:isAutoRollback rolledBackReleaseHash:stableReleaseHash errorString:errorString];
            [stateManager syncStallionMeta];
            break;

        default:
            // Default slot, no rollback needed
            break;
    }
}

+ (void)fallbackProd {
    StallionStateManager *stateManager = [StallionStateManager sharedInstance];
    NSString *baseFolderPath = stateManager.stallionConfig.filesDirectory;

  [StallionFileManager deleteFileOrFolderSilently:[NSString stringWithFormat:@"%@%@%@", baseFolderPath, StallionObjConstants.prod_directory, StallionObjConstants.new_folder_slot]];
  [StallionFileManager deleteFileOrFolderSilently:[NSString stringWithFormat:@"%@%@%@", baseFolderPath, StallionObjConstants.prod_directory, StallionObjConstants.stable_folder_slot]];
  [StallionFileManager deleteFileOrFolderSilently:[NSString stringWithFormat:@"%@%@%@", baseFolderPath, StallionObjConstants.prod_directory, StallionObjConstants.temp_folder_slot]];

    [stateManager clearStallionMeta];
}

+ (void)rollbackStage {
    StallionStateManager *stateManager = [StallionStateManager sharedInstance];
    StallionMeta *meta = stateManager.stallionMeta;

    [meta setCurrentStageSlot:SlotStateDefaultSlot];
    [meta setStageNewHash:@""];
    [stateManager syncStallionMeta];
}

+ (void)stabilizeProd {
    @try {
        StallionStateManager *stateManager = [StallionStateManager sharedInstance];
        NSString *baseFolderPath = stateManager.stallionConfig.filesDirectory;

      NSString *newSlotPath = [NSString stringWithFormat:@"%@%@%@", baseFolderPath, StallionObjConstants.prod_directory, StallionObjConstants.new_folder_slot];
      NSString *stableSlotPath = [NSString stringWithFormat:@"%@%@%@", baseFolderPath, StallionObjConstants.prod_directory, StallionObjConstants.stable_folder_slot];

      [StallionFileManager copyFileOrDirectoryFrom:newSlotPath to:stableSlotPath];

        NSString *newReleaseHash = stateManager.stallionMeta.prodNewHash;
        [stateManager.stallionMeta setProdStableHash:newReleaseHash];
        [stateManager syncStallionMeta];
        [self emitStabilizeEvent:newReleaseHash];
    } @catch (NSException *exception) {
        NSLog(@"StabilizeProd failed: %@", exception.reason);
    }
}

#pragma mark - Event Emission

+ (void)emitRollbackEvent:(BOOL)isAutoRollback rolledBackReleaseHash:(NSString *)rolledBackReleaseHash errorString:(NSString *)errorString {
    NSMutableDictionary *eventPayload = [NSMutableDictionary dictionary];
    eventPayload[@"releaseHash"] = rolledBackReleaseHash;
    eventPayload[@"error"] = errorString;

  NSString *eventName = isAutoRollback ? StallionObjConstants.auto_rolled_back_prod_event : StallionObjConstants.rolled_back_prod_event;

    if (isAutoRollback) {
        StallionStateManager *stateManager = [StallionStateManager sharedInstance];
        [stateManager.stallionMeta setLastRolledBackHash:rolledBackReleaseHash];
        [stateManager syncStallionMeta];
    }

    [[StallionEventHandler sharedInstance] sendEvent:eventName eventPayload:eventPayload];
}

+ (void)emitStabilizeEvent:(NSString *)newReleaseHash {
    NSMutableDictionary *eventPayload = [NSMutableDictionary dictionary];
    eventPayload[@"releaseHash"] = newReleaseHash;

  [[StallionEventHandler sharedInstance] sendEvent:StallionObjConstants.stabilized_prod_event eventPayload:eventPayload];
}

@end

File: android/src/main/java/com/stallion/utils/StallionSlotManager.java
package com.stallion.utils;

import com.stallion.events.StallionEventManager;
import com.stallion.storage.StallionConfigConstants;
import com.stallion.storage.StallionMetaConstants;
import com.stallion.storage.StallionStateManager;
import com.stallion.events.StallionEventConstants.NativeProdEventTypes;

import org.json.JSONObject;

import java.io.File;

public class StallionSlotManager {

  private static final StallionStateManager stateManager = StallionStateManager.getInstance();
  private static final String baseFolderPath = stateManager.getStallionConfig().getFilesDirectory();

  public static void rollbackProd(boolean isAutoRollback, String errorString) {
    StallionMetaConstants.SlotStates currentProdSlot = stateManager.stallionMeta.getCurrentProdSlot();
    String stableReleaseHash = stateManager.stallionMeta.getProdStableHash();
    String newReleaseHash = stateManager.stallionMeta.getProdNewHash();

    switch (currentProdSlot) {
      case NEW_SLOT:
        stateManager.stallionMeta.setProdNewHash("");
        if (stableReleaseHash.isEmpty()) {
          stateManager.stallionMeta.setCurrentProdSlot(StallionMetaConstants.SlotStates.DEFAULT_SLOT);
        } else {
          stateManager.stallionMeta.setCurrentProdSlot(StallionMetaConstants.SlotStates.STABLE_SLOT);
        }
        emitRollbackEvent(isAutoRollback, newReleaseHash, errorString);
        stateManager.syncStallionMeta();
        break;

      case STABLE_SLOT:
        stateManager.stallionMeta.setProdStableHash("");
        stateManager.stallionMeta.setCurrentProdSlot(StallionMetaConstants.SlotStates.DEFAULT_SLOT);
        emitRollbackEvent(isAutoRollback, stableReleaseHash, errorString);
        stateManager.syncStallionMeta();
        break;

      default:
        // Default slot, no rollback needed
        break;
    }
  }

  public static void fallbackProd() {
    StallionFileManager.deleteFileOrFolderSilently(new File(baseFolderPath + StallionConfigConstants.PROD_DIRECTORY + StallionConfigConstants.NEW_FOLDER_SLOT));
    StallionFileManager.deleteFileOrFolderSilently(new File(baseFolderPath + StallionConfigConstants.PROD_DIRECTORY + StallionConfigConstants.STABLE_FOLDER_SLOT));
    StallionFileManager.deleteFileOrFolderSilently(new File(baseFolderPath + StallionConfigConstants.PROD_DIRECTORY + StallionConfigConstants.TEMP_FOLDER_SLOT));
    stateManager.clearStallionMeta();
  }

  public static void rollbackStage() {
    stateManager.stallionMeta.setCurrentStageSlot(StallionMetaConstants.SlotStates.DEFAULT_SLOT);
    stateManager.stallionMeta.setStageNewHash("");
    stateManager.syncStallionMeta();
  }

  public static void stabilizeProd() {
    try {
      File newSlot = new File(baseFolderPath, StallionConfigConstants.PROD_DIRECTORY + StallionConfigConstants.NEW_FOLDER_SLOT);
      File stableSlot = new File(baseFolderPath, StallionConfigConstants.PROD_DIRECTORY + StallionConfigConstants.STABLE_FOLDER_SLOT);

      StallionFileManager.copyDirectory(newSlot, stableSlot);

      String newReleaseHash = stateManager.stallionMeta.getProdNewHash();
      stateManager.stallionMeta.setProdStableHash(newReleaseHash);
      stateManager.syncStallionMeta();
      emitStabilizeEvent(newReleaseHash);
    } catch (Exception ignored) {}
  }

  private static void emitRollbackEvent(boolean isAutoRollback, String rolledBackReleaseHash, String errorString) {
    try {
      JSONObject eventPayload = new JSONObject();
      eventPayload.put("releaseHash", rolledBackReleaseHash);
      eventPayload.put("meta", errorString);

      String eventName = isAutoRollback
        ? NativeProdEventTypes.AUTO_ROLLED_BACK_PROD.toString()
        : NativeProdEventTypes.ROLLED_BACK_PROD.toString();

      if (isAutoRollback) {
        stateManager.stallionMeta.setLastRolledBackHash(rolledBackReleaseHash);
        stateManager.syncStallionMeta();
      }

      StallionEventManager.getInstance().sendEvent(eventName, eventPayload);
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  private static void emitStabilizeEvent(String newReleaseHash) {
    try {
      JSONObject eventPayload = new JSONObject();
      eventPayload.put("releaseHash", newReleaseHash);

      StallionEventManager.getInstance().sendEvent(
        NativeProdEventTypes.STABILIZED_PROD.toString(),
        eventPayload
      );
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}

